internal/npm_install/npm_install.bzl

```diff
@@ -298,7 +310,51 @@ data attribute.
     ),
 })
 
-def _apply_patches(repository_ctx, patches):
+def _write_file(repository_ctx, path, contents):
+    bash_exe = repository_ctx.os.environ["BAZEL_SH"] if "BAZEL_SH" in repository_ctx.os.environ else "bash"
+    st = repository_ctx.execute(
+        [bash_exe, "-c", "echo '%s' > %s" % (contents, path)],
+        quiet = repository_ctx.attr.quiet,
+    )
+    if st.return_code:
+        fail("Error writing to file %s:\n%s%s" % (path, st.stderr, st.stdout))
+
+def _is_file(repository_ctx, path):
+    bash_exe = repository_ctx.os.environ["BAZEL_SH"] if "BAZEL_SH" in repository_ctx.os.environ else "bash"
+    st = repository_ctx.execute(
+        [bash_exe, "-c", "[ -f \"%s\" ] && echo 1 || echo 0" % path],
+        quiet = repository_ctx.attr.quiet,
+    )
+    if st.return_code:
+        fail("Error checking for existance of file %s:\n%s%s" % (path, st.stderr, st.stdout))
+    return True if st.stdout[0] == "1" else False
+
+def _apply_pre_install_patches(repository_ctx):
+    if len(repository_ctx.attr.pre_install_patches) == 0:
+        return
+    if repository_ctx.attr.symlink_node_modules:
+        fail("pre_install_patches cannot be used with symlink_node_modules enabled")
+    _apply_patches(repository_ctx, _WORKSPACE_REROOTED_PATH, repository_ctx.attr.pre_install_patches)
+
+def _apply_post_install_patches(repository_ctx):
+    if len(repository_ctx.attr.post_install_patches) == 0:
+        return
+    if repository_ctx.attr.symlink_node_modules:
+        print("\nWARNING: @%s post_install_patches with symlink_node_modules enabled will run in your workspace and potentially modify source files" % repository_ctx.name)
+    working_directory = _workspace_root(repository_ctx) if repository_ctx.attr.symlink_node_modules else _WORKSPACE_REROOTED_PATH
+
+    # marker file used when repository_ctx.attr.symlink_node_modules is True
+    marker_file = "%s/node_modules/.bazel-post-install-patches" % repository_ctx.path(repository_ctx.attr.package_json).dirname
+    allow_failure = True if repository_ctx.attr.symlink_node_modules and _is_file(repository_ctx, marker_file) else False
```

Comment:
since there's a bash dependency anyway, would it simplify the logic a bit to have a larger bash program for patches, with less starlark code? In particular maybe you don't need the write_file and is_file helpers in starlark since then you have the boilerplate of calling multiple subcommands

Reply:
good idea. done.